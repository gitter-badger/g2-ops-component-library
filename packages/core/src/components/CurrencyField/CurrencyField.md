CurrencyField:

Uncontrolled Example:

Useful when you want to have the component retaining it's own state
```js

const companyCodeMapper = require('./countryMapper').default;
const countryCode = 'DE';

<div style={{ maxWidth: '300px' }}>
  <CurrencyField
    label={'Currency Field'}
    disabled={false}
    countryCode={countryCode}
    placeholder={companyCodeMapper(countryCode).placeHolder}
    maxValue={999999999.99}
  />
</div>
```

Controlled Example:

Value set from parent as well as being set from within the component.

OnBlur triggers the call to  update the parent state/Redux
```js

const companyCodeMapper = require('./countryMapper').default;
const countryCode = 'DE';
initialState = { value: 1234.56 };

<div style={{ maxWidth: '300px' }}>
  <CurrencyField
    label={'Currency Field'}
    disabled={false}
    countryCode={countryCode}
    placeholder={companyCodeMapper(countryCode).placeHolder}
    maxValue={999999999.99}
    value={state.value}
    onBlur={(formattedValue, rawValue) => {
      console.log(formattedValue, rawValue)
      setState({ value: rawValue })
    }}
  />
</div>
```
