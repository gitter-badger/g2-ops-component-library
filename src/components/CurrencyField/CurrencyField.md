CurrencyField Example:

```js
const companyCodeMapper = require('./countryMapper').default;
const countryCode = 'DE';

<div style={{ maxWidth: '300px' }}>
  <CurrencyField
    label={'Currency Field'}
    disabled={false}
    countryCode={countryCode}
    placeholder={companyCodeMapper(countryCode).placeHolder}
    maxValue={999999999.99}
  />
</div>
```
