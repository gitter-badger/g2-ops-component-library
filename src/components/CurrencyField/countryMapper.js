// @flow

const countryCode = 'DE'

export const currencyDisplayMap = (currency: string) =>
  ({
    USD: 'symbol',
    CAD: 'code',
    GBP: 'symbol',
    EUR: 'symbol',
    AED: 'symbol',
    OMR: 'symbol',
    BHD: 'symbol',
  }[currency])

const companyCodeMapper = (country: string) =>
  ({
    US: {
      companyCode: 'COPART',
      locale: 'en-US',
      distanceUnit: 'mi',
      format: 'MM/DD/YYYY',
      region: 'America/Los_Angeles',
      currency: 'USD',
      name: 'USA',
      placeHolder: '$00,000,000.00',
      delimiter: '.',
    },
    UK: {
      companyCode: 'COPARTUK',
      locale: 'en-GB',
      distanceUnit: 'km',
      format: 'DD/MM/YYYY',
      region: 'Europe/Belfast',
      currency: 'GBP',
      name: 'UK',
      placeHolder: '£00,000,000.00',
      delimiter: '.',
    },
    IR: {
      // need more clarity on the details for this country
      companyCode: 'COPARTUK',
      locale: 'en-GB',
      distanceUnit: 'km',
      format: 'DD/MM/YYYY',
      region: 'Europe/Belfast',
      currency: 'EUR',
      name: 'IR',
      placeHolder: '€00,000,000.00',
      delimiter: '.',
    },
    ME: {
      // need more clarity on the details for this country
      companyCode: 'COPARTUK',
      locale: 'en-GB',
      distanceUnit: 'km',
      format: 'DD/MM/YYYY',
      region: 'Europe/Belfast',
      currency: 'AED',
      name: 'MEA',
      placeHolder: 'AED00,000,000.00',
      delimiter: '.',
    },
    CA: {
      companyCode: 'COPART',
      locale: 'en-US',
      distanceUnit: 'km',
      region: 'America/Los_Angeles',
      format: 'MM/DD/YYYY',
      currency: 'CAD',
      name: 'CANADA',
      placeHolder: 'CAD00,000,000.00',
      delimiter: '.',
    },
    DE: {
      companyCode: 'COPART',
      locale: 'de-DE',
      distanceUnit: 'km',
      format: 'DD/MM/YYYY',
      region: 'Europe/Berlin',
      currency: 'EUR',
      name: 'GERMANY',
      placeHolder: '00.000.000,00 €',
      delimiter: ',',
    },
    ES: {
      companyCode: 'COPART',
      locale: 'de-DE',
      distanceUnit: 'km',
      format: 'DD/MM/YYYY',
      region: 'Europe/Belfast',
      currency: 'EUR',
      name: 'SPAIN',
      placeHolder: '00.000.000,00 €',
      delimiter: ',',
    },
    IN: {
      companyCode: 'COPART',
      locale: 'en-IN',
      distanceUnit: 'km',
      region: 'Asia/Kolkata',
      format: 'MM/DD/YYYY',
      currency: 'INR',
      name: 'INDIA',
      placeHolder: '₹ 00,00,000.00',
      delimiter: '.',
    },
    GB: {
      companyCode: 'COPARTUK',
      locale: 'en-GB',
      distanceUnit: 'km',
      format: 'DD/MM/YYYY',
      region: 'Europe/Belfast',
      currency: 'GBP',
      name: 'GB',
      placeHolder: '£00,000,000.00',
      delimiter: '.',
    },
  }[country || countryCode])

export default companyCodeMapper
