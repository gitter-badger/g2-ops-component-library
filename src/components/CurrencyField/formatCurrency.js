export const serialize = (countryCode, inputValue) => {
  switch (countryCode.toUpperCase()) {
    case 'IN':
      return parseFloat(inputValue.replace(/[^\d.]/g, ''))
    case 'DE':
    case 'ES': {
      const commasToPoints = inputValue.replace(/\D|\./g, (matched) => (matched === ',' ? '.' : ''))
      return parseFloat(commasToPoints)
    }
    default:
  }
}
const isDigit = (char) => char.match(/^\d$/)
export const getFormatterRTL = (TEMPLATE) => (intermediaryValueM, charsDiff, intermediarySelectionStartM) => {
  const isNegative = intermediaryValueM.startsWith('-')
  const intermediaryValue = isNegative ? intermediaryValueM.slice(1) : intermediaryValueM
  const intermediarySelectionStart = intermediarySelectionStartM + (isNegative ? 1 : 0)
  const templateCharDeleted = TEMPLATE[intermediarySelectionStart] !== '@'
  const withoutDeletedDigits =
    charsDiff === -1 && templateCharDeleted // what was deleted was a template character
      ? intermediaryValue.slice(0, intermediarySelectionStart).replace(/\d\D@$/, '') + // remove template characters before these
        intermediaryValue.slice(intermediarySelectionStart)
      : intermediaryValue

  if (withoutDeletedDigits.replace(/[^\d.,]/g, '') === '') {
    return ''
  }
  const index = (i) => TEMPLATE.length - i
  const minus = isNegative ? '-' : ''
  return (
    minus +
    [...withoutDeletedDigits.padStart(3, '0')].reduceRight((output, char) => {
      const position = index(output.length)

      let newChars
      if (output.length === TEMPLATE.length || !isDigit(char)) {
        newChars = char === TEMPLATE[position] ? char : ''
      } else {
        const nonDigitsAfter = TEMPLATE.slice(position).match(/^[^@]+/) || []
        const nonDigitsBefore = TEMPLATE.slice(0, position).match(/[^@]+$/) || []
        newChars = nonDigitsAfter + char + nonDigitsBefore
      }
      return newChars + output
    }, '')
  )
}
const formatIndianNumber = getFormatterRTL('@@,@@,@@,@@,@@@.@@')
const formatEuros = getFormatterRTL('@@@.@@@.@@@.@@@,@@ €')
const FORMATS = {
  DE: formatEuros,
  ES: formatEuros,
  IN: (...args) => {
    const result = formatIndianNumber(...args)
    return result && `₹${result}`
  },
}

export const format = (countryCode, number) => {
  const newValue = String(number).replace(/[^\d.,]/g, '')
  return FORMATS[countryCode](newValue)
}

export const formatWithNegativeNumbers = (countryCode, number) => {
  const newValue = String(number).replace(/((?!^)-|[^\d\-.,])/g, '')
  return FORMATS[countryCode](newValue)
}

export const displayFormattedCurrencyValue = (country, value = 0.0) => format(country, (value * 100).toFixed())
