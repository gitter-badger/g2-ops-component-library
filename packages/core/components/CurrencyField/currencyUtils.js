// @flow

import countryMapper, { currencyDisplayMap } from './countryMapper'

export const serialize = (countryCode: string, inputValue: string): number => {
  switch (countryCode.toUpperCase()) {
    case 'IN':
      return parseFloat(inputValue.replace(/[^\d.]/g, ''))
    case 'DE':
    case 'ES': {
      const commasToPoints = inputValue.replace(/\D|\./g, (matched) => (matched === ',' ? '.' : ''))
      return parseFloat(commasToPoints)
    }
    default:
      return 0.0
  }
}

export const getItem = (key: string): ?string => {
  const value = localStorage.getItem(key)
  return value && JSON.parse(value)
}

export const CURRENCY_STRIP_DOWN_REGEX_UK = /[^0-9.]+/g
export const CURRENCY_STRIP_DOWN_REGEX_US = /^[$-]?(\d+)[,]?(\d{0,3}\.?\d*)/
export const CURRENCY_STRIP_DOWN_REGEX = (countryCode: string): RegExp =>
  ({
    US: CURRENCY_STRIP_DOWN_REGEX_US,
    UK: CURRENCY_STRIP_DOWN_REGEX_UK,
  }[countryCode || 'US'])

export const countryLocaleMapper = (countryCode: string): string =>
  ({
    DEU: 'de-DE',
    ESP: 'de-DE',
    IND: 'en-IN',
  }[countryCode])

export const currencyReplaceValue = (countryCode: string): string =>
  ({
    US: '$1$2',
    UK: '',
  }[countryCode || 'US'])

const formatCurrency = (country: string, value: string, currencyStyle: string, clocale: ?string = null): string =>
  `${new Intl.NumberFormat(clocale || countryMapper(country).locale, {
    style: 'currency',
    currency: `${currencyStyle}`,
    currencyDisplay: `${currencyDisplayMap(currencyStyle) || 'symbol'}`,
  }).format(value)}`

export default formatCurrency

export const stripDownCurrency = (country: string, currencyValue: string = '') =>
  currencyValue.toString().replace(CURRENCY_STRIP_DOWN_REGEX(country), currencyReplaceValue(country))

/*
  This function rounds off a given currency field value to 2 declimals by default.
  2 decimals is a standard for payment fields across other applications as well.
*/
export const roundCurrency = (value: string, roundOffDecimal: number = 2) => +(+value || 0).toFixed(roundOffDecimal)
