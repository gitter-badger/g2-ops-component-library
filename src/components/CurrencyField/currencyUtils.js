import countryMapper, { currencyDisplayMap } from './countryMapper'

export const getItem = (key) => {
  const value = localStorage.getItem(key)
  return value && JSON.parse(value)
}

export const CURRENCY_STRIP_DOWN_REGEX_UK = /[^0-9\.]+/g
export const CURRENCY_STRIP_DOWN_REGEX_US = /^[\$-]?(\d+)[,]?(\d{0,3}\.?\d*)/
export const CURRENCY_STRIP_DOWN_REGEX = (countryCode) =>
  ({
    US: CURRENCY_STRIP_DOWN_REGEX_US,
    UK: CURRENCY_STRIP_DOWN_REGEX_UK,
  }[countryCode || 'US'])

export const countryLocaleMapper = (cntryCode) =>
  ({
    DEU: 'de-DE',
    ESP: 'de-DE',
    IND: 'en-IN',
  }[cntryCode])

export const currencyReplaceValue = (countryCode) => 
  ({
    US: '$1$2',
    UK: '',
  }[countryCode || 'US'])

const formatCurrency = (country, value, currencyStyle, clocale = null) =>
  `${new Intl.NumberFormat(clocale || countryMapper(country).locale, {
    style: 'currency',
    currency: `${currencyStyle}`,
    currencyDisplay: `${currencyDisplayMap(currencyStyle) || 'symbol'}`,
  }).format(value)}`

export default formatCurrency

export const zeroDollars = (currencyCode) => formatCurrency(0, currencyCode)

export const stripDownCurrency = (country, currencyValue = '') =>
  currencyValue.toString().replace(CURRENCY_STRIP_DOWN_REGEX(country), currencyReplaceValue(country))

/*
  This function rounds off a given currency field value to 2 declimals by default.
  2 decimals is a standard for payment fields across other applications as well.
*/
export const roundCurrency = (value, roundOffDecimal = 2) => +(+value || 0).toFixed(roundOffDecimal)
//
// /*
//   Returns the default currency for the present environment. Mainly used in C stack.
// */
// export const defaultCurrency = () =>
//   (countryCode === 'UK'
//     ? propOr('GBP', 'selectedCurrency', getItem('dashboard'))
//     : propOr('USD', 'currency', countryMapper(propOr(countryCode, 'selectedCountry', getItem('dashboard')))))
