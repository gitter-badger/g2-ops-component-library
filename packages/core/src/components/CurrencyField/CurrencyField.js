// @flow
import type { Node, ComponentType } from 'react'

import React, { Component } from 'react'
import { TextField } from 'components/TextField'
import { pick, equals } from 'ramda'

import formatCurrency, { stripDownCurrency, serialize } from './currencyUtils'
import companyCodeMapper from './countryMapper'

const getDelimiter = countryCode => (countryCode.toUpperCase() === 'IN' ? '.' : ',')
const isBlank = val => val == null || (typeof val === 'string' && !val.trim().length)
const CASCountry = ['US', 'UK', 'CA', 'IR', 'ME', 'GB']
const GlobalCountry = ['DE', 'ES', 'IN']
const countriesSupported = [...CASCountry, ...GlobalCountry]

type CurrencyFieldPropsT = {
  name: string,
  value: number | string,
  label: string,
  countryCode: 'US' | 'UK' | 'CA' | 'IR' | 'ME' | 'GB' | 'DE' | 'ES' | 'IN',
  disabled: boolean,
  required: boolean,
  onChange: (string, string) => void,
  onBlur: (string, string) => void,
  maxValue: number,
}

type CurrencyFieldValidationType = {
  errorMessage: string,
  formattedValue: string,
  displayedValue: string,
}

type CurrencyFieldStateT = {
  errorMessage: string,
  formattedValue: string,
  displayedValue: string,
}

const validateInputValueAndReturnErrorMessage = (
  value: string,
  countryCode: string,
  maxValue: number,
): CurrencyFieldStateT => {
  if (!countriesSupported.includes(countryCode)) {
    return {
      formattedValue: value,
      displayedValue: value,
      errorMessage: 'Please enter/send a valid country',
    }
  } else {
    const country = companyCodeMapper(countryCode)
    let currencyValue = ''
    if (!isBlank(value)) {
      currencyValue = CASCountry.includes(countryCode)
        ? stripDownCurrency(countryCode, value).replace(/[a-zA-Z,$#^&*()@!]+/, '') || 0
        : serialize(countryCode, value)
    }
    const formattedCurrency = formatCurrency(countryCode, currencyValue.toString(), country.currency)
    if (Number(currencyValue) > maxValue) {
      return {
        formattedValue: formattedCurrency,
        displayedValue: currencyValue.toString(),
        errorMessage: `Max Limit ${formatCurrency(countryCode, maxValue.toString(), country.currency)}`,
      }
    } else {
      return {
        formattedValue: formattedCurrency,
        displayedValue: currencyValue.toString(),
        errorMessage: '',
      }
    }
  }
}

const arePropValuesEqual = (
  nextProps: any,
  props: any,
  properties: Array<string> = ['maxValue', 'countryCode', 'value'],
): boolean => !equals(pick(properties, nextProps), pick(properties, props))

export class CurrencyField extends Component<CurrencyFieldPropsT, CurrencyFieldStateT> {
  static defaultProps = {
    countryCode: 'US',
    value: '',
  }

  constructor(props: CurrencyFieldPropsT) {
    super(props)
    const { value, countryCode, maxValue } = props
    this.state = validateInputValueAndReturnErrorMessage(value.toString(), countryCode, maxValue)
  }

  componentWillReceiveProps(nextProps: CurrencyFieldPropsT) {
    if (arePropValuesEqual(nextProps, this.props)) {
      const { value, countryCode, maxValue } = nextProps
      this.setState(prevState => ({
        ...prevState,
        ...validateInputValueAndReturnErrorMessage(value.toString(), countryCode, maxValue),
      }))
    }
  }

  onBlur = (e: SyntheticKeyboardEvent<HTMLInputElement>) => {
    const { countryCode, maxValue } = this.props
    const value = e.currentTarget.value
    const validatedInput = validateInputValueAndReturnErrorMessage(value, countryCode, maxValue)
    this.setState(prevState => ({
      ...prevState,
      ...validatedInput,
      displayedValue: validatedInput.formattedValue,
    }))
    if (this.props.onBlur) {
      const { formattedValue, displayedValue } = this.state
      this.props.onBlur(formattedValue, validatedInput.displayedValue.toString()) // To be consistent we return both strings
    }
  }

  onChange = (value: string) => {
    const { countryCode, maxValue } = this.props
    const validatedInput = validateInputValueAndReturnErrorMessage(value, countryCode, maxValue)
    let currencyValue = validatedInput.displayedValue
    if (countriesSupported.includes(countryCode)) {
      const country = companyCodeMapper(countryCode)
      const delimiter = country.delimiter
      currencyValue = value.replace(/\D/g, match => (match === delimiter ? delimiter : ''))
    }
    this.setState(prevState => ({
      ...prevState,
      ...validatedInput,
      displayedValue: currencyValue,
    }))
    if (this.props.onChange) {
      const { formattedValue, displayedValue } = this.state
      this.props.onChange(formattedValue, displayedValue)
    }
  }

  render() {
    const { onChange, disabled, ...restProps } = this.props
    return (
      <div>
        <TextField
          {...restProps}
          disabled={disabled}
          value={this.state.displayedValue}
          onChanged={this.onChange}
          onBlur={this.onBlur}
          onFocus={e => this.onChange(e.target.value)}
          errorMessage={this.state.errorMessage}
        />
        <input type="hidden" name={this.props.name} required={this.props.required} value={this.state.formattedValue} />
      </div>
    )
  }
}


