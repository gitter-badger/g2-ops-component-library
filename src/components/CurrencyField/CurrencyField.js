import React, { Component } from 'react'
import PropTypes from 'prop-types'
import { TextField } from 'office-ui-fabric-react/lib/TextField'
import formatCurrency, { stripDownCurrency } from './currencyUtils'
import { serialize } from './formatCurrency'
import companyCodeMapper from './countryMapper'
import { wrapMuiContext } from '../../wrapMuiContext'

const getDelimiter = countryCode => (countryCode.toUpperCase() === 'IN' ? '.' : ',')
const isBlank = (val) => val == null || (typeof val === 'string' && !val.trim().length)
const CASCountry = ['US', 'UK', 'CA', 'IR', 'ME', 'GB']
const GlobalCountry = ['DE', 'ES', 'IN']
const formatValue = (country, value, currencyStyle) => formatCurrency(
  country,
  value || 0.00,
  currencyStyle,
)

class CurrencyField extends Component {
  constructor(props) {
    super(props)
    const { value, countryCode } = props
    const country = countryCode.toUpperCase()
    
    const companyMapper = companyCodeMapper(country.toUpperCase() || 'US')
    let errorMessage = '', displayedValue = 0.00, currency = 'USD', locale = 'en-US'

    if (!(GlobalCountry.includes(country) || CASCountry.includes(country))) {
      errorMessage = 'Please enter/send a valid country'
    } else {
      currency = companyMapper && companyMapper.currency
      locale = companyMapper && companyMapper.locale
      displayedValue = isBlank(value, currency) ? '' : formatCurrency(country, value, currency, locale)
    }

    this.state = {
      errorMessage,
      value: value || 0,
      displayedValue,
      currency,
      locale,
    }
  }
  // componentWillReceiveProps(nextProps) {
  //   if (nextProps.hasOwnProperty('value') && document.activeElement !== this.textField.input) {
  //     const { locale, currency } = this.state
  //     const displayedValue = isBlank(nextProps.value) ? '' : formatCurrency(nextProps.value, currency, locale)
  //     this.setState({ value: nextProps.value, displayedValue })
  //   }
  // }

  onBlur = e => {
    const input = e.target.value
    const { currency, locale } = this.state
    const { countryCode, maxValue } = this.props
    const country = countryCode.toUpperCase()

    const inputFieldIsBlank = isBlank(input)
    let replacedValue = ''

    if (CASCountry.includes(country)) {
      replacedValue = stripDownCurrency(country, input).replace(/[a-zA-Z,$#\^&*\(\)@!]+/, '') || 0
      
      if (replacedValue && replacedValue !== '' && replacedValue !== '0.00' && replacedValue !== '.') {
        if (+replacedValue <= maxValue) {
          this.setState({ displayedValue: formatValue(country, replacedValue, currency), value: replacedValue })
        } else {
          this.setState({ displayedValue: formatValue(country, maxValue, currency), value: replacedValue })
        }
      } else {
        this.setState({ displayedValue: formatValue(country, 0, currency), value: replacedValue })
      }
      if (this.props.onBlur) {
        this.props.onBlur(e)
      }
    } else {
      replacedValue = inputFieldIsBlank ? '' : Number(serialize(country, input)).toFixed(2)
      
      this.setState({
        displayedValue: inputFieldIsBlank ? null : formatCurrency(country, replacedValue, currency, locale),
        value: replacedValue,
      })
      if (this.props.onChange) {
        this.props.onChange(e, replacedValue)
      }
    }
  }

  handleChange = input => {
    const { countryCode, maxValue } = this.props
    const country = countryCode.toUpperCase()

    if (!(GlobalCountry.includes(country) || CASCountry.includes(country))) {
      this.setState({ errorMessage: 'Not a valid country' })
      return
    }
    const delimiter = getDelimiter(country)
    const displayedValue = CASCountry.includes(country)
      ? stripDownCurrency(country, input)
      : input.replace(/\D/g, (match) => (match === delimiter ? delimiter : ''))

    this.setState({ displayedValue })

    if (Number(displayedValue) <= this.props.maxValue) {
      this.setState({ errorMessage: '' })
      this.setState({ displayedValue })
      // calling handle change from props if its there
      if (this.props.handleChange) {
        this.props.handleChange(event, displayedValue)
      }
    } else {
      this.setState({ errorMessage: `Max Limit $${maxValue}` })
    }
  }

  serialize = value => serialize(this.props.countryCode, value)

  render() {
    const {
      onChange,
      ...restProps
    } = this.props
    return (
      <div>
        <TextField
          {...restProps}
          ref={elem => {this.textField = elem}}
          value={this.state.displayedValue}
          onChanged={this.handleChange}
          onBlur={this.onBlur}
          onFocus={(e) => this.handleChange(e.target.value)}
          errorMessage={this.state.errorMessage}
        />
        <input type="hidden"
          name={this.props.name}
          required={this.props.required}
          value={this.state.value}
          />
      </div>
    )
  }
}
CurrencyField.propTypes = {
  name: PropTypes.string,
  value: PropTypes.oneOfType([PropTypes.number, PropTypes.string]),
  label: PropTypes.string,
  countryCode: PropTypes.string.isRequired,
  disabled: PropTypes.bool,
  required: PropTypes.bool,
  onChange: PropTypes.func,
}
export default wrapMuiContext(CurrencyField)