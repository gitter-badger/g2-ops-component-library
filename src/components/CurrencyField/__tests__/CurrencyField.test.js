import React from 'react'
import renderer from 'react-test-renderer'
import { mount } from 'enzyme'

import CurrencyField from '../CurrencyField'
import companyCodeMapper from '../countryMapper'

describe('<CurrencyField />', () => {
  test('should render properly', () => {
    const countries = [ 'US', 'UK', 'IR', 'ME', 'CA', 'DE', 'ES', 'IN', 'GB' ]
    countries.forEach((countryCode) => {
      const tree = renderer.create(
        <CurrencyField
          label={'Currency Field'}
          disabled={false}
          countryCode={countryCode}
          placeholder={companyCodeMapper(countryCode).placeHolder}
          maxValue={99.99}
        />,
      )
      expect(tree).toMatchSnapshot(tree)
    })
  })

  test('should throw error when maxValue is reached', () => {
    const countryCodeErrorMap = {
      DE: 'Max Limit € 99.99',
      US: 'Max Limit $99.99',
      UK: 'Max Limit £99.99',
      IR: 'Max Limit €99.99',
      CA: 'Max Limit CAD99.99',
      IN: 'Max Limit ₹99.99',
      GB: 'Max Limit £99.99',
    }
    Object.keys(countryCodeErrorMap).forEach((countryCode) => {
      const tree = mount(
        <CurrencyField
          label={'Currency Field'}
          disabled={false}
          countryCode={countryCode}
          placeholder={companyCodeMapper(countryCode).placeHolder}
          maxValue={99.99}
        />,
      )
      tree
        .find('TextField')
        .at(0)
        .prop('onChanged')('100.00')
      expect(tree.state('errorMessage')).toBe(countryCodeErrorMap[countryCode])
    })
  })

  test.skip('should return formatted currency value when entered value < maxValue', () => {
    const countryCodeErrorMap = {
      DE: 'Max Limit € 99.99',
      US: 'Max Limit $99.99',
      UK: 'Max Limit £99.99',
      IR: 'Max Limit €99.99',
      CA: 'Max Limit CAD99.99',
      IN: 'Max Limit ₹99.99',
      GB: 'Max Limit £99.99',
    }
    Object.keys(countryCodeErrorMap).forEach((countryCode) => {
      const onBlur = jest.fn()
      const tree = mount(
        <CurrencyField
          label={'Currency Field'}
          disabled={false}
          countryCode={countryCode}
          placeholder={companyCodeMapper(countryCode).placeHolder}
          maxValue={999999999.99}
          onBlur={onBlur}
        />,
      )
      tree
        .find('TextField')
        .at(0)
        .prop('onBlur')({
          currentTarget: {
            value: '10000000.00',
          },
        })
      expect(onBlur).toHaveBeenCalledWith('')
    })
  })
})
