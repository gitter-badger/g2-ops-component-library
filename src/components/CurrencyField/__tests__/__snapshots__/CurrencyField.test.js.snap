// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CurrencyField /> should render properly: [object Object] 1`] = `
<div>
  <div
    className="ms-Fabric css-32"
  >
    <div
      className="ms-TextField root"
    >
      <div
        className="ms-TextField-wrapper wrapper"
      >
        <label
          className="ms-Label root-33"
          htmlFor="TextField0"
        >
          Currency Field
        </label>
        <div
          className="ms-TextField-fieldGroup fieldGroup"
        >
          <input
            aria-describedby={null}
            aria-invalid={false}
            aria-label={undefined}
            className="ms-TextField-field field input "
            disabled={false}
            id="TextField0"
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            onInput={[Function]}
            placeholder="$00,000,000.00"
            type="text"
            value=""
          />
        </div>
      </div>
    </div>
  </div>
  <input
    name={undefined}
    required={undefined}
    type="hidden"
    value={undefined}
  />
</div>
`;

exports[`<CurrencyField /> should render properly: [object Object] 2`] = `
<div>
  <div
    className="ms-Fabric css-32"
  >
    <div
      className="ms-TextField root"
    >
      <div
        className="ms-TextField-wrapper wrapper"
      >
        <label
          className="ms-Label root-33"
          htmlFor="TextField2"
        >
          Currency Field
        </label>
        <div
          className="ms-TextField-fieldGroup fieldGroup"
        >
          <input
            aria-describedby={null}
            aria-invalid={false}
            aria-label={undefined}
            className="ms-TextField-field field input "
            disabled={false}
            id="TextField2"
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            onInput={[Function]}
            placeholder="£00,000,000.00"
            type="text"
            value=""
          />
        </div>
      </div>
    </div>
  </div>
  <input
    name={undefined}
    required={undefined}
    type="hidden"
    value={undefined}
  />
</div>
`;

exports[`<CurrencyField /> should render properly: [object Object] 3`] = `
<div>
  <div
    className="ms-Fabric css-32"
  >
    <div
      className="ms-TextField root"
    >
      <div
        className="ms-TextField-wrapper wrapper"
      >
        <label
          className="ms-Label root-33"
          htmlFor="TextField4"
        >
          Currency Field
        </label>
        <div
          className="ms-TextField-fieldGroup fieldGroup"
        >
          <input
            aria-describedby={null}
            aria-invalid={false}
            aria-label={undefined}
            className="ms-TextField-field field input "
            disabled={false}
            id="TextField4"
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            onInput={[Function]}
            placeholder="€00,000,000.00"
            type="text"
            value=""
          />
        </div>
      </div>
    </div>
  </div>
  <input
    name={undefined}
    required={undefined}
    type="hidden"
    value={undefined}
  />
</div>
`;

exports[`<CurrencyField /> should render properly: [object Object] 4`] = `
<div>
  <div
    className="ms-Fabric css-32"
  >
    <div
      className="ms-TextField root"
    >
      <div
        className="ms-TextField-wrapper wrapper"
      >
        <label
          className="ms-Label root-33"
          htmlFor="TextField6"
        >
          Currency Field
        </label>
        <div
          className="ms-TextField-fieldGroup fieldGroup"
        >
          <input
            aria-describedby={null}
            aria-invalid={false}
            aria-label={undefined}
            className="ms-TextField-field field input "
            disabled={false}
            id="TextField6"
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            onInput={[Function]}
            placeholder="AED00,000,000.00"
            type="text"
            value=""
          />
        </div>
      </div>
    </div>
  </div>
  <input
    name={undefined}
    required={undefined}
    type="hidden"
    value={undefined}
  />
</div>
`;

exports[`<CurrencyField /> should render properly: [object Object] 5`] = `
<div>
  <div
    className="ms-Fabric css-32"
  >
    <div
      className="ms-TextField root"
    >
      <div
        className="ms-TextField-wrapper wrapper"
      >
        <label
          className="ms-Label root-33"
          htmlFor="TextField8"
        >
          Currency Field
        </label>
        <div
          className="ms-TextField-fieldGroup fieldGroup"
        >
          <input
            aria-describedby={null}
            aria-invalid={false}
            aria-label={undefined}
            className="ms-TextField-field field input "
            disabled={false}
            id="TextField8"
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            onInput={[Function]}
            placeholder="CAD00,000,000.00"
            type="text"
            value=""
          />
        </div>
      </div>
    </div>
  </div>
  <input
    name={undefined}
    required={undefined}
    type="hidden"
    value={undefined}
  />
</div>
`;

exports[`<CurrencyField /> should render properly: [object Object] 6`] = `
<div>
  <div
    className="ms-Fabric css-32"
  >
    <div
      className="ms-TextField root"
    >
      <div
        className="ms-TextField-wrapper wrapper"
      >
        <label
          className="ms-Label root-33"
          htmlFor="TextField10"
        >
          Currency Field
        </label>
        <div
          className="ms-TextField-fieldGroup fieldGroup"
        >
          <input
            aria-describedby={null}
            aria-invalid={false}
            aria-label={undefined}
            className="ms-TextField-field field input "
            disabled={false}
            id="TextField10"
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            onInput={[Function]}
            placeholder="00.000.000,00 €"
            type="text"
            value=""
          />
        </div>
      </div>
    </div>
  </div>
  <input
    name={undefined}
    required={undefined}
    type="hidden"
    value={undefined}
  />
</div>
`;

exports[`<CurrencyField /> should render properly: [object Object] 7`] = `
<div>
  <div
    className="ms-Fabric css-32"
  >
    <div
      className="ms-TextField root"
    >
      <div
        className="ms-TextField-wrapper wrapper"
      >
        <label
          className="ms-Label root-33"
          htmlFor="TextField12"
        >
          Currency Field
        </label>
        <div
          className="ms-TextField-fieldGroup fieldGroup"
        >
          <input
            aria-describedby={null}
            aria-invalid={false}
            aria-label={undefined}
            className="ms-TextField-field field input "
            disabled={false}
            id="TextField12"
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            onInput={[Function]}
            placeholder="00.000.000,00 €"
            type="text"
            value=""
          />
        </div>
      </div>
    </div>
  </div>
  <input
    name={undefined}
    required={undefined}
    type="hidden"
    value={undefined}
  />
</div>
`;

exports[`<CurrencyField /> should render properly: [object Object] 8`] = `
<div>
  <div
    className="ms-Fabric css-32"
  >
    <div
      className="ms-TextField root"
    >
      <div
        className="ms-TextField-wrapper wrapper"
      >
        <label
          className="ms-Label root-33"
          htmlFor="TextField14"
        >
          Currency Field
        </label>
        <div
          className="ms-TextField-fieldGroup fieldGroup"
        >
          <input
            aria-describedby={null}
            aria-invalid={false}
            aria-label={undefined}
            className="ms-TextField-field field input "
            disabled={false}
            id="TextField14"
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            onInput={[Function]}
            placeholder="₹ 00,00,000.00"
            type="text"
            value=""
          />
        </div>
      </div>
    </div>
  </div>
  <input
    name={undefined}
    required={undefined}
    type="hidden"
    value={undefined}
  />
</div>
`;

exports[`<CurrencyField /> should render properly: [object Object] 9`] = `
<div>
  <div
    className="ms-Fabric css-32"
  >
    <div
      className="ms-TextField root"
    >
      <div
        className="ms-TextField-wrapper wrapper"
      >
        <label
          className="ms-Label root-33"
          htmlFor="TextField16"
        >
          Currency Field
        </label>
        <div
          className="ms-TextField-fieldGroup fieldGroup"
        >
          <input
            aria-describedby={null}
            aria-invalid={false}
            aria-label={undefined}
            className="ms-TextField-field field input "
            disabled={false}
            id="TextField16"
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            onInput={[Function]}
            placeholder="£00,000,000.00"
            type="text"
            value=""
          />
        </div>
      </div>
    </div>
  </div>
  <input
    name={undefined}
    required={undefined}
    type="hidden"
    value={undefined}
  />
</div>
`;
